### PR TITLE
Fixes read metadata failed after dropped partition for V1 format

### LINKED ISSUE #3374: select .files and rewrite data files failed after remove partition field
if update the spec, remove the partition field, then cannot query files, and rewrite data files, expire snapshots will also raise exception

### _Case_
```
create table iceberg_catalog.db.bucket2(a string, b string, c string) using iceberg partitioned by (bucket(2, b), bucket(2, c)) location "/Users/jcc/hive/bucket2";
insert into db.bucket2 values("1", "cc", "bb");
insert into db.bucket2 values("1", "cc", "bb");
alter table iceberg_catalog.db.bucket2 drop partition field bucket(2, c);
insert into db.bucket2 values("1", "cc", "bb");
insert into db.bucket2 values("1", "cc", "bb");
select * from iceberg_catalog.db.bucket2.files;
```
### _Exception is_
```
Caused by: java.lang.IllegalArgumentException: Wrong class, java.lang.CharSequence, for object: 1
        at org.apache.iceberg.PartitionData.get(PartitionData.java:120)
        at org.apache.iceberg.spark.source.StructInternalRow.getUTF8StringInternal(StructInternalRow.java:175)
        at org.apache.iceberg.spark.source.StructInternalRow.getUTF8String(StructInternalRow.java:171)
        at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
        at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
        at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:729)
        at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:340)
        at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:872)
        at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:872)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:349)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:313)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
        at org.apache.spark.scheduler.Task.run(Task.scala:127)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by ConeyLiu**
This patch fixes two problems:
1. For V1 tables, we use a `VoidTransform` to replace the removed partition field. While the result type of `VoidTransform` same as the type of the field. For example, the result type of `Bucket(2, string_type_field)` is int, while is string for `VoidTransform(string_type_field)`. So we should use the original partition field type(int instead of string) to build the common partitioning type.
2. We should use the matched `PartitionSpec` to convert the `PartitionFiedlSummary` to human string, not the current table spec.

Closes #3374

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 1]
```diff
@@ -228,11 +229,22 @@ public static StructType partitionType(Table table) {
           fieldMap.put(fieldId, field);
           NestedField structField = spec.partitionType().field(fieldId);
           structFields.add(structField);
+
```
[the remark this answers is not in the document]
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:232]: Do we need the code changes in the other files then? Seems like the only bug is here for V1 tables with columns transformed from a functional partition spec to a void spec with a different type?

**Comment by RussellSpitzer** [comment]
OK I think I understand the issue now, this is specifically an error where the attempt to build a common Partition Spec builds a spec using void transforms in V1 tables, the void column has a different return type than the original partition transform and throws an error. Is that correct?

I would recommend renaming this issue/pr to match the underlying issue. I also am guessing because this is the issue, there is no problem with V2 tables.

@aokolnychyi Should probably also take a look

File: core/src/main/java/org/apache/iceberg/AllManifestsTable.java [hunk 2]
```diff
@@ -134,14 +136,14 @@ public long targetSplitSize() {
...
-          return new ManifestListReadTask(ops.io(), schema(), table().spec(), new BaseFileScanTask(
+          return new ManifestListReadTask(ops.io(), schema(), specs, new BaseFileScanTask(
               manifestListAsDataFile, null,
               schemaString, specString, residuals));
         } else {
           return StaticDataTask.of(
               ops.io().newInputFile(ops.current().metadataFileLocation()),
               MANIFEST_FILE_SCHEMA, schema(), snap.allManifests(),
-              manifest -> ManifestsTable.manifestFileToRow(table().spec(), manifest)
+              manifest -> ManifestsTable.manifestFileToRow(specs.get(manifest.partitionSpecId()), manifest)
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/AllManifestsTable.java:146]: Is this change needed, shouldn't we be using the common spec from table built in Partitioning?

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 3]
```diff
@@ -63,6 +67,28 @@ private void preparePartitionedTable() {
         .commit();
   }
 
+  @Test
+  public void testManifestsTableWithDroppedPartition() throws IOException {
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:71]: I think the key part of this test is we should be checking a dropped V1 Field whose return type changes when dropped. As is this test I believe would always pass on a V2 table even without this patch.

**Comment by aokolnychyi** [comment]
I may be able to take a look within the next few days

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 4]
```diff
@@ -63,6 +67,28 @@ private void preparePartitionedTable() {
...
+        .appendFile(FILE_A)
+        .appendFile(FILE_B)
+        .commit();
+
+    table.updateSpec().removeField(Expressions.bucket("data", 16)).commit();
+    table.refresh();
+
+    Table manifestsTable = new ManifestsTable(table.ops(), table);
+    TableScan scan = manifestsTable.newScan()
+            .filter(Expressions.lessThan("length", 10000L));
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:82]: Why are we filtering?

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 5]
```diff
@@ -228,11 +229,22 @@ public static StructType partitionType(Table table) {
...
+          if (Transforms.alwaysNull().equals(field.transform())) {
+            fieldToIndex.put(fieldId, structFields.size() - 1);
+          }
         } else {
           // verify the fields are compatible as they may conflict in v1 tables
           ValidationException.check(equivalentIgnoringNames(field, existingField),
               "Conflicting partition fields: ['%s', '%s']",
               field, existingField);
+
+          if (Transforms.alwaysNull().equals(existingField.transform()) &&
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:242]: I think we should add a comment here,

"Void Transforms always revert back to the Type of the original column being transformed. This will lead to type errors when analyzing partition fields written for earlier files with a different type. To avoid this we always restore NullTransforms to their previous non-null state."

Or something like that?

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 6]
```diff
@@ -228,11 +229,22 @@ public static StructType partitionType(Table table) {
...
+                  !Transforms.alwaysNull().equals(field.transform())) {
+            fieldMap.put(fieldId, field);
+            NestedField structField = spec.partitionType().field(fieldId);
+            structFields.set(fieldToIndex.get(fieldId), structField);
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:246]: I think this may be easier (and we wouldn't need to build up the index map) to just do
```java
fieldMap.put(fieldId, field);
NestedField previousTransform  = spec.partitionType().field(fieldId);
voidTransformIndex = structFields.indexOf(existingField)
structFields.set(voidTransformIndex, previousTransform)
```

I recommend this because I think it more clearly expresses what we are trying to do here. When initially looked at the code It wasn't clear to me that the fieldToIndex was solely for looking up old void transforms for this edge case.

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 7]
```diff
@@ -63,6 +67,28 @@ private void preparePartitionedTable() {
         .commit();
   }
 
+  @Test
+  public void testManifestsTableWithDroppedPartition() throws IOException {
+    table.newFastAppend()
+        .appendFile(FILE_A)
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:73]: I think for completeness sake we should also have tests for the other Partition reading metadata tables. (Files, Manfiests, ManifestEntries) or what not

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 8]
```diff
@@ -140,9 +140,10 @@ public void testPartitionTypeWithAddingBackSamePartitionFieldInV1Table() {
         .addField("data")
         .commit();
 
-    // in v1, we use void transforms instead of dropping partition fields
+    // in v1, we use void transforms instead of dropping partition fields.
+    // We restored the void transforms with the original dropped partition fields.
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:144]: I'm not sure line 144 here makes sense to me, I think this is fine without the comment. In fact I think the expected type makes more sense now than it previously did.

File: core/src/main/java/org/apache/iceberg/AllManifestsTable.java (hunk 2 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/AllManifestsTable.java:146]: `Partitioning.partitionType` builds a common partition type, here we need a `PartitionSpec` argument.

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java (hunk 3 above)
Comment (ConeyLiu) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:71]: It should be only in V1. I think it should be OK that the tests covered v2 as well, that's right?

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java (hunk 7 above)
Comment (ConeyLiu) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:73]: Should the following tests(`testAllManifestsTableWithDroppedPartition` and `testFilesTableScanWithDroppedPartition`) have covered the case ?

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java (hunk 4 above)
Comment (ConeyLiu) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:82]: It could be removed. The test is adjusted from `testManifestsTableAlwaysIgnoresResiduals`.

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java (hunk 8 above)
Comment (ConeyLiu) [reply core/src/test/java/org/apache/iceberg/TestPartitioning.java:144]: OK, remove it.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 6 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:246]:
```java
fieldMap.put(fieldId, field); 
NestedField previousTransform = spec.partitionType().field(fieldId); 
int voidTransformIndex = structFields.indexOf(existingField);  // here, we need a `NestedField` instance not a `PartitionField` instance, and it seems that we could not get the previous `PartitionField` because we are in new spec.
structFields.set(voidTransformIndex, previousTransform)```

Please see the above inline comments.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:232]: We should update it, otherwise we could get the error linked here: https://github.com/apache/iceberg/issues/3374#issuecomment-951841026
```

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 5 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:245]: added

**Comment by ConeyLiu** [comment]
Thanks @RussellSpitzer for the review.

> OK I think I understand the issue now, this is specifically an error where the attempt to build a common Partition Spec builds a spec using void transforms in V1 tables, the void column has a different return type than the original partition transform and throws an error. Is that correct?

Yes, that's right.

> I would recommend renaming this issue/pr to match the underlying issue. I also am guessing because this is the issue, there is no problem with V2 tables.

Any suggestion for this?  What about `Fixes read metadata failed after dropped partition in V1 tables`?

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java (hunk 7 above)
Comment (RussellSpitzer) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:76]: ManfiestEntries and partitions should be slightly different right?

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java (hunk 3 above)
Comment (RussellSpitzer) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:74]: Yeah I just wanted to make sure we note that this test was added to check the transformation of a partition column's type. Like we are actually checking something a little more specific that dropping a partition, since if "data" was an int there wouldn't be a bug.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 6 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:249]: ok got it, we need the "spec.partitionType()" of the spec the partition field was created with, and at this point we have lost that reference

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 9]
```diff
@@ -211,6 +211,7 @@ public static StructType partitionType(Table table) {
 
     Map<Integer, PartitionField> fieldMap = Maps.newHashMap();
     List<NestedField> structFields = Lists.newArrayList();
+    Map<Integer, Integer> fieldToIndex = Maps.newHashMap();
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:214]: I still think this needs a new name ... since we only use it to track indexes of void transformed fields

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 10]
```diff
@@ -66,6 +70,28 @@ private void preparePartitionedTable() {
...
+        .appendFile(FILE_B)
+        .commit();
+
+    table.updateSpec().removeField(Expressions.bucket("data", 16)).commit();
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:80]: Could we add a test which drops and adds partitioning more than once? I just want to have confidence for something like

Int -> Void<String> ->  Int -> Void<String> -> Truncate<String>

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 6 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/Partitioning.java]: Yeah, that's right.

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java (hunk 7 above)
Comment (ConeyLiu) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:76]: Added

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 9 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/Partitioning.java]: Updated the name, and left comments here.

**Comment by ConeyLiu** [comment]
@RussellSpitzer, I am sorry for the late update. Thanks for the review. Comments have been addressed. Pls take another look when you are free.

**Comment by aokolnychyi** [comment]
Sorry for the delay on my side. This PR is the next one I'll look into. I took a quick look and I think I remember the context.

**Comment by aokolnychyi** [comment]
Okay, I think I got the issue. Essentially, we don't know the correct type produced by Void transforms that we assign in v1 tables when a partition field is dropped. We assume the output type matches the source field type but it is not always the case. Let me take a look at the solution on Monday.

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 11]
```diff
@@ -140,9 +140,9 @@ public void testPartitionTypeWithAddingBackSamePartitionFieldInV1Table() {
         .addField("data")
         .commit();
 
-    // in v1, we use void transforms instead of dropping partition fields
+    // in v1, we use void transforms instead of dropping partition fields.
     StructType expectedType = StructType.of(
-        NestedField.optional(1000, "data_1000", Types.StringType.get()),
+        NestedField.optional(1000, "data", Types.StringType.get()),
```
Comment (aokolnychyi) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:145]: This change is wrong. The previous expected value was correct as we need to inherit the newest column names.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 12]
```diff
@@ -211,6 +211,8 @@ public static StructType partitionType(Table table) {
 
     Map<Integer, PartitionField> fieldMap = Maps.newHashMap();
     List<NestedField> structFields = Lists.newArrayList();
+    // Mapping from a PartitionField with void transform to the index in structFields.
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:214]: It is a little bit hard to follow these changes and they also break the logic for inheriting updated column names.

What about something like this?

```
Map<Integer, PartitionField> fieldMap = Maps.newHashMap();
Map<Integer, Type> typeMap = Maps.newHashMap();
Map<Integer, String> nameMap = Maps.newHashMap();

// sort the spec IDs in descending order to pick up the most recent field names
List<Integer> specIds = table.specs().keySet().stream()
    .sorted(Collections.reverseOrder())
    .collect(Collectors.toList());

for (Integer specId : specIds) {
  PartitionSpec spec = table.specs().get(specId);

  for (PartitionField field : spec.fields()) {
    int fieldId = field.fieldId();
    NestedField structField = spec.partitionType().field(fieldId);

    PartitionField existingField = fieldMap.get(fieldId);

    if (existingField == null) {
      fieldMap.put(fieldId, field);
      typeMap.put(fieldId, structField.type());
      nameMap.put(fieldId, structField.name());

    } else {
      // verify the fields are compatible as they may conflict in v1 tables
      ValidationException.check(equivalentIgnoringNames(field, existingField),
          "Conflicting partition fields: ['%s', '%s']",
          field, existingField);

      // use the correct type for dropped partitions in v1 tables
      if (isVoidTransform(existingField) && !isVoidTransform(field)) {
        fieldMap.put(fieldId, field);
        typeMap.put(fieldId, structField.type());
      }
    }
  }
}

List<NestedField> sortedStructFields = fieldMap.keySet().stream()
    .sorted(Comparator.naturalOrder())
    .map(fieldId -> NestedField.optional(fieldId, nameMap.get(fieldId), typeMap.get(fieldId)))
    .collect(Collectors.toList());
return StructType.of(sortedStructFields);
```

```
private static boolean isVoidTransform(PartitionField field) {
  return field.transform().equals(Transforms.alwaysNull());
}
```

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java (hunk 11 above)
Comment (ConeyLiu) [reply core/src/test/java/org/apache/iceberg/TestPartitioning.java]: reverted

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 12 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/Partitioning.java]: Updated. Thanks for the suggestion.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 13]
```diff
@@ -210,7 +211,8 @@ public static StructType partitionType(Table table) {
     }
 
     Map<Integer, PartitionField> fieldMap = Maps.newHashMap();
-    List<NestedField> structFields = Lists.newArrayList();
+    Map<Integer, Type> typeMap = Maps.newHashMap();
+    Map<Integer, String> nameMap = Maps.newHashMap();
```
Comment (szlta) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:215]: I have run into similar issues, and I think this will help resolve the type change of columns in older specs.
Another thing I have seen and is probably still a problem, is that this method may return a column name multiple times. Consider the following:
Table schema: a int, b date, c date
spec0: year(b), a
spec1: a
spec2: year(b), year(c)
then the result is something like this: 1000: b_year int, 1001: a int, 1002: b_year int, 1003: c_year int
Further down the line when we construct a Schema object, we will have a failure due to b_year name being present in two fields (1000, 1002).
How should this case be handled? Maybe appending _r<fieldId> to each column name?
cc: @RussellSpitzer , @aokolnychyi

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: @szlta, Iceberg should rename older columns when it updates the spec to avoid conflicts. That's why we have to take the latest column names. I think we should have some tests in `TestPartitioning`. Could you check, @szlta?

Comment (szlta) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: @aokolnychyi  I think you're referring to what's happening in V1 tables. For those the spec is ever-growing in a way that no partition fields/transforms are removed, but rather converted to void.
The rename logic is there: https://github.com/apache/iceberg/blob/master/core%2Fsrc%2Fmain%2Fjava%2Forg%2Fapache%2Ficeberg%2FBaseUpdatePartitionSpec.java#L179
but I think this is not used for V2 tables, as per https://github.com/apache/iceberg/blob/master/core%2Fsrc%2Fmain%2Fjava%2Forg%2Fapache%2Ficeberg%2FBaseUpdatePartitionSpec.java#L261-L268

Since in V2, specs don't retain old deleted partition fields this rename is not required for normal operations.

The problem I'm describing only affects the metadata table queries, because for V2, due to the lack of above renames, Partitioning.partitionType() collects all partition fields from all previous specs too. With the lack of renames this can result in the same field name being present multiple times, and cause the PartitionsTable's (or DataFilesTable's) schema to be failed to get constructed.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: I guess you are right, @szlta. Here is an example to reproduce.

```
PartitionSpec initialSpec = PartitionSpec.builderFor(SCHEMA)
    .identity("data")
    .build();
TestTables.TestTable table = TestTables.create(tableDir, "test", SCHEMA, initialSpec, V2_FORMAT_VERSION);

table.updateSpec()
    .removeField("data")
    .commit();

table.updateSpec()
    .addField("data")
    .addField("id")
    .commit();

struct<1000: data: optional string, 1001: data: optional string, 1002: id: optional int>
```

While it would be great to update the logic that evolves the spec, I think we have to adapt the method that builds a common representation too. Maintaining a set of used names and appending a suffix of the field ID sounds like a reasonable approach.

Thoughts, @rdblue @RussellSpitzer @szehon-ho @flyrain?

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: Looks like we are using field IDs while projecting values from the common representation in Spark.
At least, that part is working.

```
protected Map<Integer, StructProjection> buildPartitionProjections(Types.StructType partitionType,
                                                                   Map<Integer, PartitionSpec> specs) {
  Map<Integer, StructProjection> partitionProjections = Maps.newHashMap();
  specs.forEach((specID, spec) ->
      partitionProjections.put(specID, StructProjection.create(partitionType, spec.partitionType()))
  );
  return partitionProjections;
}
```

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: I think Field_id Suffix seems very reasonable to me.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: I think we have two cases to solve this problem. The combined partition struct type is used in both metadata tables and when we need to project the _partition field for certain queries. For queries that use the _partition field, I think adding a suffix to make the names unique is the right approach. That's internal so it doesn't really matter what we rename to, as long as we get the projections that produce the partition tuple for a given spec right.

For metadata tables, we expect the field names to match the partition names. A simple example is that we use identity partitions using the original column name. So it would be weird to partition by `category` and need to query `category_r1003` in the metadata table. I think that the right thing to do for metadata tables is to have a separate way to produce the combined struct that produces only one partition column per name.

In @aokolnychyi's example, `1001: data` should be present and `1000: data` should not be used for metadata tables.

This is also an area where we may want to bring back compatible partition columns. There's no reason why Anton's example couldn't detect that `1000: data` was in an old spec and reuse the ID to avoid this problem.

**Comment by rdblue** [comment]
@ConeyLiu, can you update the description? It is really helpful to not only say what you're trying to fix, but also to describe the approach and why the changes are needed. Thanks!

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 13 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: I'm also up for solution 2, where we rename just the columns from older specs. That sounds like a reasonable solution.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: I'll catch up today. Sorry for the delay!

**Comment by ConeyLiu** [comment]
> @ConeyLiu, can you update the description? It is really helpful to not only say what you're trying to fix, but also to describe the approach and why the changes are needed. Thanks!

Updated.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 14]
```diff
@@ -222,27 +224,41 @@ public static StructType partitionType(Table table) {
 
       for (PartitionField field : spec.fields()) {
         int fieldId = field.fieldId();
+        NestedField structField = spec.partitionType().field(fieldId);
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:228]: Nit: no need for a newline between assignment statements.

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 15]
```diff
@@ -140,7 +140,7 @@ public void testPartitionTypeWithAddingBackSamePartitionFieldInV1Table() {
         .addField("data")
         .commit();
 
-    // in v1, we use void transforms instead of dropping partition fields
+    // in v1, we use void transforms instead of dropping partition fields.
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:143]: Please remove this change since it is non-functional. Also, we don't require punctuation to end comments.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 16]
```diff
@@ -222,27 +224,41 @@ public static StructType partitionType(Table table) {
...
+          }
         }
       }
     }
 
-    List<NestedField> sortedStructFields = structFields.stream()
-        .sorted(Comparator.comparingInt(NestedField::fieldId))
+    List<NestedField> sortedStructFields = fieldMap.keySet().stream()
+        .sorted(Comparator.naturalOrder())
+        .map(fieldId -> NestedField.optional(fieldId, nameMap.get(fieldId), typeMap.get(fieldId)))
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:253]: I think the changes in this file are correct and fix the first issue, where the type may be incorrect for void transform fields in v1 tables.

File: core/src/main/java/org/apache/iceberg/ManifestsTable.java [hunk 17]
```diff
@@ -73,10 +71,15 @@ MetadataTableType metadataTableType() {
   protected DataTask task(TableScan scan) {
     TableOperations ops = operations();
     String location = scan.snapshot().manifestListLocation();
+    Map<Integer, PartitionSpec> specs = table().specs();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/ManifestsTable.java:74]: Can you update this to copy the map into a new `HashMap`? Not all maps are serializable with Kryo, so we should make sure it is created by `Maps.newHashMap` to be safe.

File: core/src/main/java/org/apache/iceberg/AllManifestsTable.java [hunk 18]
```diff
@@ -118,6 +119,7 @@ protected String tableType() {
         boolean ignoreResiduals, boolean caseSensitive, boolean colStats) {
       String schemaString = SchemaParser.toJson(schema());
       String specString = PartitionSpecParser.toJson(PartitionSpec.unpartitioned());
+      Map<Integer, PartitionSpec> specs = table().specs();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/AllManifestsTable.java:122]: Similar to the `ManifestsTable`, I think this needs to copy the specs into a `newHashMap`.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 13 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:215]: I think that the fixes in this PR look good other than a couple minor updates that are needed. We should definitely follow up with a PR that fixes the names as suggested by @szlta.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 14 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/Partitioning.java]: removed

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java (hunk 15 above)
Comment (ConeyLiu) [reply core/src/test/java/org/apache/iceberg/TestPartitioning.java]: reverted

File: core/src/main/java/org/apache/iceberg/ManifestsTable.java (hunk 17 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/ManifestsTable.java]: updated

File: core/src/main/java/org/apache/iceberg/AllManifestsTable.java (hunk 18 above)
Comment (ConeyLiu) [reply core/src/main/java/org/apache/iceberg/AllManifestsTable.java]: updated

File: core/src/main/java/org/apache/iceberg/ManifestsTable.java [hunk 19]
```diff
@@ -73,10 +72,15 @@ MetadataTableType metadataTableType() {
...
     TableOperations ops = operations();
     String location = scan.snapshot().manifestListLocation();
+    Map<Integer, PartitionSpec> specs = Maps.newHashMap(table().specs());
+
     return StaticDataTask.of(
         ops.io().newInputFile(location != null ? location : ops.current().metadataFileLocation()),
         schema(), scan.schema(), scan.snapshot().allManifests(),
-        manifest -> ManifestsTable.manifestFileToRow(spec, manifest)
+        manifest -> {
+          PartitionSpec partitionSpec = specs.get(manifest.partitionSpecId());
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ManifestsTable.java:81]: nit: just `spec`?

**Review by aokolnychyi — APPROVED** [review]
This looks ready to me. Could you rebase it, @ConeyLiu? There seems to be a merge conflict.

We should follow up with what @szlta brought up. There is #4292 to track that.

**Comment by ConeyLiu** [comment]
Rebased, thanks for the reminder. @aokolnychyi

**Comment by RussellSpitzer** [comment]
Thanks @ConeyLiu !

**Review by RussellSpitzer — APPROVED** [review]

**Comment by RussellSpitzer** [comment]
Also thanks to all of the reviewers who took their time to check on this PR, I know this was a tricky issue and I'm glad we had so many eyes on the issue.

**Comment by ConeyLiu** [comment]
Thanks all!

**Comment by szehon-ho** [comment]
Should this fix be included in 0.13.2?

**Comment by rdblue** [comment]
@szehon-ho, that sounds fine to me.

**Comment by szehon-ho** [comment]
:thumbsup: looks like it has popped up in a few places.  

@ConeyLiu  let me know if you want to make a pr against 0.13.x branch and I can review, otherwise i will take a look.  (not sure the process exactly)

**Comment by ConeyLiu** [comment]
Thanks @szehon-ho, will do it.
